File: examples/class-kits/gospider/ecommerce_review.go
package classkit

import scrapyapi "gospider/scrapy"

func NewEcommerceReviewSpider() *scrapyapi.Spider {
	profile := profileForFamily(defaultSiteFamily)
	return scrapyapi.NewSpider("ecommerce-review", func(response *scrapyapi.Response) ([]any, error) {
		family := siteFamilyFromResponse(response.URL, nil)
		if response.Request != nil {
			family = siteFamilyFromResponse(response.URL, response.Request.Meta)
		}
		current := profileForFamily(family)

		return []any{
			scrapyapi.NewItem().
				Set("kind", "ecommerce_review").
				Set("site_family", family).
				Set("url", response.URL).
				Set("item_id", firstRegexMatch(response.Text, current.ItemIDPatterns)).
				Set("rating", firstRegexMatch(response.Text, current.RatingPatterns)).
				Set("review_count", firstRegexMatch(response.Text, current.ReviewCountPatterns)).
				Set("review_id_candidates", collectRegexMatches(response.Text, []string{`(?:commentId|reviewId|id)["'=:\s]+([A-Za-z0-9_-]+)`}, 10)).
				Set("excerpt", excerpt(response.Text, 800)).
				Set("note", "Template for public review pages or review APIs. Prefer stable JSON payloads over brittle DOM selectors."),
		}, nil
	}).
		AddStartURL(profile.ReviewURL).
		WithStartMeta("site_family", defaultSiteFamily).
		WithStartMeta("runner", profile.Runner)
}

File: examples/class-kits/rustspider/ecommerce_review.rs
use std::sync::Arc;

use rustspider::scrapy::project as projectruntime;
use rustspider::scrapy::{Item, Output, Response, Spider};

use super::ecommerce_profile::{
    collect_matches, first_match, profile_for_family, site_family_from_response, text_excerpt,
    DEFAULT_SITE_FAMILY,
};

pub fn make_ecommerce_review_spider() -> Spider {
    let profile = profile_for_family(DEFAULT_SITE_FAMILY);
    Spider::new(
        "ecommerce-review",
        Arc::new(|response: &Response| {
            let family = site_family_from_response(response);
            let current = profile_for_family(&family);
            vec![Output::Item(
                Item::new()
                    .set("kind", "ecommerce_review")
                    .set("site_family", family)
                    .set("url", response.url.clone())
                    .set("item_id", first_match(&response.text, current.item_id_patterns))
                    .set("rating", first_match(&response.text, current.rating_patterns))
                    .set(
                        "review_count",
                        first_match(&response.text, current.review_count_patterns),
                    )
                    .set(
                        "review_id_candidates",
                        collect_matches(
                            &response.text,
                            &[r#"(?:commentId|reviewId|id)["'=:\s]+([A-Za-z0-9_-]+)"#],
                            10,
                        ),
                    )
                    .set("excerpt", text_excerpt(&response.text, 800))
                    .set(
                        "note",
                        "Template for public review pages or review APIs. Prefer stable JSON payloads over brittle DOM selectors.",
                    ),
            )]
        }),
    )
    .add_start_url(profile.review_url)
    .with_start_meta("site_family", DEFAULT_SITE_FAMILY)
    .with_start_meta("runner", profile.runner)
}

pub fn register() {
    projectruntime::register_spider("ecommerce-review", make_ecommerce_review_spider);
}
